File: src/pkg08_parameters_cpp/src/params_node.cpp
// params_node.cpp
// Demonstrates declaring, reading, and reacting to parameter changes.

#include <memory>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

class ParamsNode : public rclcpp::Node
{
public:
  ParamsNode() : Node("params_node")
  {
    // 1) Declare parameters with default values.
    //    Anything not declared first cannot be set later (since Foxy).
    this->declare_parameter<std::string>("robot_name", "turtle");
    this->declare_parameter<int>("max_speed", 10);
    this->declare_parameter<double>("loop_rate_hz", 1.0);
    this->declare_parameter<std::vector<std::string>>(
      "waypoints", std::vector<std::string>{"A", "B", "C"});

    // 2) Read parameters
    auto robot_name = this->get_parameter("robot_name").as_string();
    auto max_speed = this->get_parameter("max_speed").as_int();
    auto rate = this->get_parameter("loop_rate_hz").as_double();
    auto waypoints = this->get_parameter("waypoints").as_string_array();

    RCLCPP_INFO(get_logger(),
      "robot_name=%s, max_speed=%ld, loop_rate_hz=%.2f",
      robot_name.c_str(), max_speed, rate);

    for (const auto & wp : waypoints) {
      RCLCPP_INFO(get_logger(), "  waypoint: %s", wp.c_str());
    }

    // 3) React to parameter changes at runtime
    param_callback_handle_ = this->add_on_set_parameters_callback(
      std::bind(&ParamsNode::on_param_change, this, std::placeholders::_1));

    // Timer to show current value periodically
    timer_ = create_wall_timer(2s, [this]() {
      RCLCPP_INFO(get_logger(), "Current max_speed = %ld",
        this->get_parameter("max_speed").as_int());
    });
  }

private:
  rcl_interfaces::msg::SetParametersResult on_param_change(
    const std::vector<rclcpp::Parameter> & params)
  {
    rcl_interfaces::msg::SetParametersResult result;
    result.successful = true;

    for (const auto & p : params) {
      if (p.get_name() == "max_speed") {
        if (p.as_int() < 0 || p.as_int() > 100) {
          result.successful = false;
          result.reason = "max_speed must be in [0, 100]";
          RCLCPP_WARN(get_logger(), "Rejected: %s", result.reason.c_str());
        } else {
          RCLCPP_INFO(get_logger(), "max_speed updated to %ld", p.as_int());
        }
      }
    }
    return result;
  }

  OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ParamsNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/pkg09_parameters_py/pkg09_parameters_py/params_node.py
#!/usr/bin/env python3
"""Demonstrates declaring, reading, and reacting to parameters in Python."""

import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult


class ParamsNode(Node):
    def __init__(self):
        super().__init__('params_node_py')

        # 1. Declare with default values
        self.declare_parameter('robot_name', 'turtle')
        self.declare_parameter('max_speed', 10)
        self.declare_parameter('loop_rate_hz', 1.0)
        self.declare_parameter('waypoints', ['A', 'B', 'C'])

        # 2. Read at startup
        robot_name = self.get_parameter('robot_name').value
        max_speed = self.get_parameter('max_speed').value
        rate = self.get_parameter('loop_rate_hz').value
        waypoints = self.get_parameter('waypoints').value

        self.get_logger().info(
            f'robot_name={robot_name}, max_speed={max_speed}, '
            f'loop_rate_hz={rate}')
        for wp in waypoints:
            self.get_logger().info(f'  waypoint: {wp}')

        # 3. React to runtime changes
        self.add_on_set_parameters_callback(self.on_param_change)

        # Show current value every 2s
        self.timer = self.create_timer(2.0, self.print_current)

    def on_param_change(self, params):
        result = SetParametersResult(successful=True)
        for p in params:
            if p.name == 'max_speed':
                if p.value < 0 or p.value > 100:
                    result.successful = False
                    result.reason = 'max_speed must be in [0, 100]'
                    self.get_logger().warn(f'Rejected: {result.reason}')
                else:
                    self.get_logger().info(f'max_speed updated to {p.value}')
        return result

    def print_current(self):
        self.get_logger().info(
            f'Current max_speed = {self.get_parameter("max_speed").value}')


def main(args=None):
    rclpy.init(args=args)
    node = ParamsNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
